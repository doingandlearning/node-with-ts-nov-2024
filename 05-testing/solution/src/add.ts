export default function add(a: number, b: number) {
  if (typeof a != "number" || typeof b != "number") {
    throw new TypeError("You need to pass in numbers.");
  }
  return a + b;
}
