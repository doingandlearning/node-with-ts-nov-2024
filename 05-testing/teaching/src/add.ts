export function add(a: number, b: number) {
  if (typeof a !== "number" || typeof b !== "number") {
    throw new TypeError("Add function must only be called with numbers");
  }
  return a + b;
}
