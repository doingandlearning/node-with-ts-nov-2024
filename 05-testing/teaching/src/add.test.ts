import { test, expect, describe } from "vitest";
import { add } from "./add";

test("the add function correctly adds two numbers", () => {
  // Arrange Act Assert
  // Given When Then
  const testCases = [
    [2, 3, 5],
    [-1, 0, -1],
    [9, 10, 19],
    [0.1, 0.3, 0.4],
  ]; // [a, b, result]

  // parameterizing our test  (@pytest.mark.parameterize())
  for (let testCase of testCases) {
    const result = add(testCase[0], testCase[1]);
    expect(result).toEqual(testCase[2]);
  }
});

// def test_throws_an_error_if_called_with
describe("unhappy path for the add function", () => {
  test("throws an error if called with non-number arguments", () => {
    expect(() => add(true as any, [] as any)).toThrow();
    expect(() =>
      add(true as any, [] as any)
    ).toThrowErrorMatchingInlineSnapshot(
      `[TypeError: Add function must only be called with numbers]`
    );
    expect(() => add(true as any, [] as any)).toThrowErrorMatchingSnapshot();
  });
});
