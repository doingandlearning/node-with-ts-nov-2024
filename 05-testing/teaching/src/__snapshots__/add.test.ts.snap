// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unhappy path for the add function > throws an error if called with non-number arguments 2`] = `[TypeError: Add function must only be called with numbers]`;
